src/operator/subgraph/subgraph_property.h

```diff
@@ -296,8 +296,20 @@ class SubgraphProperty {
    */
   virtual void ConnectSubgraphOutputs(const nnvm::NodePtr subgraph_node,
                                       std::vector<nnvm::NodeEntry*>* output_entries) const {
+    // Collapse output_entries pointing to same NodeEntry
```

Comment:
Can we implement this outside of ConnectSubgraphOutputs? In other words, can we ensure that `output_entries` in `ConnectSubgraphOutputs` doesn't have duplicated entries, and we can create mapping in build_subgraph to achieve that.

Reply:
Any suggestion on how to do this without creating a map or changing the interface/API of createSubgraphOutputs?

Reply:
It is important to be clear about 2 things :
1. "output_entries" are misnamed. They are really input_entries_consuming_subgraph_output.
2. Therefore, output_entries do not have duplicate entries. The number of output_entries is correct ! (or rather the the number of input_entries_consuming_subgraph_output is correct)

Consider the following 3 node json example -
(json simplified to eliminate irrelevant fields)

```
nodes : [ 
  { name : "NodeA" }, #nodes[0]
  { name : "NodeB",   #nodes[1]
    inputs : [[0,0,0]] }, #pointing at NodeA 
  { name : "NodeC",  #nodes[2]
    inputs : [[0,0,0]] } #pointing at NodeA 
]
```

Suppose NodeA gets turned into a subgraph with 1 node. If we do that then, 
The so called "output_entries" are :
1. nodes[1][inputs][0] #0th input of NodeB
2. nodes[2][inputs[0] #0th input of NodeC

Therefore the variable names "output_entries" is actually refering to the the "input_entries_consuming_subgraph_output". 
The correct number of "output_entries" is therefore 2. 
We can not collapse this down to 1 because both NodeB and NodeC need to point back to NodeA

The real outputs are "sym.outputs" in the build_graph.cc code in the PR. 
We do de-duplicate those.

Reply:
Of course each entries have their unique address, which means, they are presenting different data consuming for this node's output. But, if some entries are the same(same node, index and version), then they should still be the 'same' after subgraph partition(same means the entry value is the same among those entries, but of course the entry value is different from before partition).
Take your example as example,
After partition, the result would be
```
nodes : [ 
  { name : "SubgraphA" }, #subgraph_nodes[0], have 2 outputs, the output[1] presents the origin output of  NodeA
  { name : "NodeB",   #nodes[1]
    inputs : [[0,1,0]] }, #pointing at SubgraphA, repsents the output data require of NodeA 
  { name : "NodeC",  #nodes[2]
    inputs : [[0,1,0]] } #pointing at SubgraphA, repsents the output data require of NodeA 
]
```
So, here's what we can do to simplify the work in `ConnectSubgraphOutputs` 

1. Create a map about unique_entry -> origin entry address(Not necessary, just for better performance).
2. Pass unique_entry to `ConnectSubgraphOutputs`, and return a map of  unique_entry -> unique_entry_after_partition.
3. Then we know, how each original entry should be replaced for. Then do the replacement on all entry address based on above 2 maps.

What I what to do here is, we should avoid backend developer to handle the duplicated entries, instead, backend should only tells the new connection after partition. Take above case as example, backend developer should only tells that the output data requirement on NodeA now should pointing to index 1 on new subgraph node through `ConnectSubgraphOutputs`. And build_subgraph pass should propagate this change on all original entries which points at NodeA.

Reply:
@zhennanqin are you suggesting we change the API for connect subgraph outputs? That would be a breaking change with lots of implications. 

Are you ok with doing this in 1.6 timeframe? Maybe we should take @hahtk suggestion for now and change the API for the 2.0 release?

Reply:
Don't need to change API. The pseudo code is,
```
std::vector<nnvm::NodeEntry*> unique_output_entires; // Collect the unique entry, let's collect the first entry
std::unordered_map<nnvm::NodeEntry*, std::vector<nnvm::NodeEntry*>> output_entires_map; // Collect all duplicated
// Collect above infomation
for (auto entry_ptr : output_entries) {
  if (*entry_ptr is unique) { // replace this with nnvm::NodeEntryEqual logic
    unique_output_entires.push_back(entry_ptr);
  } else {
    output_entires_map[first_entry_of_it].push_back(entry_ptr);
  }
}
// Pass unique_output_entires to ConnectSubgraphOutputs and collect the change
subg_prop->ConnectSubgraphOutputs(n, &unique_output_entires);
// Now we know how to make the change to all duplicated
for (auto entry_ptr : unique_output_entires) {
  auto duplicated_entries = output_entires_map[entry_ptr];
  for (auto& i : duplicated_entries) {
    *i = *entry_ptr;
  }
}
```

Reply:
You're right. But changing inside ConnectSubgraphOutputs() won't help for the backends that override this function, and because of duplicated output entries, most of backends had already overrided this function. Your proposal doesn't work for them. Even for newly added backends, they will most likely continue overriding ConnectSubgraphOutputs() because the outputs connection may have a different sequence. 

To make things more clear as you requested, I suggest to introduce an new api, maybe called `ConnectSubgraphUniqueOutputs`, then the code would be like,
```
virtual void ConnectSubgraphUniqueOutputs(const nnvm::NodePtr subgraph_node,
                                          std::vector<nnvm::NodeEntry*>* output_entries) const {
  size_t idx = 0;
  for (size_t i = 0; i < output_entries->size(); ++i) {
    *output_entries->at(i) = nnvm::NodeEntry{subgraph_node, idx++, 0};
}

virtual void ConnectSubgraphOutputs(const nnvm::NodePtr subgraph_node,
                                      std::vector<nnvm::NodeEntry*>* output_entries) const {
std::vector<nnvm::NodeEntry*> unique_output_entires; // Collect the unique entry, let's collect the first entry
std::unordered_map<nnvm::NodeEntry*, std::vector<nnvm::NodeEntry*>> output_entires_map; // Collect all duplicated
// Collect above infomation
for (auto entry_ptr : output_entries) {
  if (*entry_ptr is unique) { // replace this with nnvm::NodeEntryEqual logic
    unique_output_entires.push_back(entry_ptr);
  } else {
    output_entires_map[first_entry_ptr_of_it].push_back(entry_ptr);
  }
}
// Pass unique_output_entires to ConnectSubgraphUniqueOutputs and collect the change
ConnectSubgraphUniqueOutputs(n, &unique_output_entires);
// Now we know how to make the change to all duplicated
for (auto entry_ptr : unique_output_entires) {
  auto duplicated_entries = output_entires_map[entry_ptr];
  for (auto& i : duplicated_entries) {
    *i = *entry_ptr;
  }
}
```
Then backend developer can only override ConnectSubgraphUniqueOutputs() to avoid handle duplicated outputs.

Reply:
Hey Guys,

I think we're getting offtrack here. Lets just clarify what we're working toward. We want the subgraph to only have unique outputs, so this code is necessary in creating the sym.outputs correctly:

https://github.com/apache/incubator-mxnet/pull/16131/files#diff-962e4e83b570943c087e477058485e25R578-R587

Then, the problem is we need to wire up the those unique outputs to each entry in output_entries. Since we pass the new symbol "sym" into connectSubgraphOutputs indirectly via the node "n", technically the set of unique outputs from the subgraph is already available in the connectSubgraphOutputs function. 

To clarify, the ideal solution would be to pass in a map from output node to output entry and then let connectSubgraphOutputs just wire them up as it sees fit. 

Do we agree that we do not want to change the API of connectSubgraphOutputs to enable this? Is there any reason not to just do this now? it shouldnt be too much work to go change, there is only a few subgraph properties that would need to change:

https://github.com/apache/incubator-mxnet/search?q=connectSubgraphOutputs&unscoped_q=connectSubgraphOutputs

And I think the creators of those wouldnt mind just fixing this once and for all now. This would go in the 1.6 release.

Reply:
@samskalicky Before this PR, backend still has the chance to remove the duplicated sym.outputs in CreateSubgraphNode(). The existing APIs are enough to solve the problem you're trying to fix, so I think this PR is to make all of this happen in a automatic way. But the change in connectSubgraphOutputs() doesn't achieve that goal, but making the solution semi-automatic. 

From a backend developer position, I rather prefer to have full control of my code(current APIs works in this way), or making most of the code automatically(the way I suggest to change). For the semi-automatic solution, developer have to spend extra effort to figure out which parts are handled automatically(sym.outputs), and which parts are not(connectSubgraphOutputs). And also increased the change that backend developer to make a bug when overriding connectSubgraphOutputs(I can't find any reason for connecting entries with same value to different sym.outputs, but by mistake).

Anyway, I don't against to make the change in your way. At least maybe we should add some comments in connectSubgraphOutputs to highlight the difference between sym.outputs and output_entries, and let developer notice that they need to handle the duplicated entires carefully?

Reply:
@ZhennanQin Would you approve changing the connectSubgraphOutputs API to pass in a map instead for output_entries? Then it will simplify the implementation of connectSubgraphOutputs and help backend developers avoid making bugs by having to dedupe in this weird way. 

@HahTK and I are willing to make this change. 

Heres what we propose:

Change FindOutputEntries to populate a map<int, vector<nodeentry*> > instead of the vector<nodeentry*>. The integer refers to the i'th output of the subgraph, and nodeentry pointer is the same as it is in output_entries. The job of connectSubgraphOutputs is to connect all nodeentries in the vector to that indexed output of the subgraph.

Then in CreateSubgraphNode (in build_subgraph.cc) the code will look like:

```
nnvm::Symbol sym;
  size_t idx = 0;
  sym.outputs.resize(output_map.size());
  for (size_t i = 0; i < output_map.size(); ++i) {
    sym.outputs[i] = *output_map[i][0];
  }
```

And ConnectSubgraphOutputs will take the output_map and do:

```
for (size_t i = 0; i < output_map->size(); ++i) {
   for(auto e : output_map[i]) {
     *e = nnvm::NodeEntry{subgraph_node, static_cast<uint32_t>(i), 0};
    }
}
```

This vastly simplifies/clarifies the code backend devs need to write. And helps avoid confusion and prevents bugs.

Reply:
This solution is more clear! I like this change. Please go ahead to implement it. Thanks.